openedx/openedx-platform
# Detailed Permissions for Content Libraries

This PR introduces more fine-grained permissions for blockstore-based content libraries and uses Bridgekeeper to enforce them.

Any individual user or group can be added to a content library with one of these three permissions levels:
* View only
* Author (edit XBlocks and the library)
* Admin (All author permissions, plus edit user roles and delete the library)

In addition there are two flags that can be set at the library level:
* Allow public access (anyone can view the library content including OLX source)
* Allow public learning (anyone can learn from the XBlocks in the course, i.e. view them directly and submit answers; when this is false, the XBlocks can only be used if they are added to a course or another library that does allow public learning)

Permissions can be easily modified using the REST API or the django admin:

<img width="1103" alt="Screen Shot 2019-12-11 at 7 00 42 PM" src="https://user-images.githubusercontent.com/945577/70678978-974f4880-1c48-11ea-9dd0-b66d7c3948d4.png">

Test instructions:

Use [Ramshackle](https://github.com/open-craft/ramshackle#ramshackle) to create a content library with some content. Use the django admin to grant a subset of permissions (like author or read only) to another user, login to Studio as that other user, then use ramshackle and verify that you get errors when trying to do prohibited actions.

And/or manually run the new test cases:

Start a blockstore test server (`make testserver` in the `blockstore` dir), then from `make studio-shell`, run:

```
EDXAPP_RUN_BLOCKSTORE_TESTS=1 python -Wd -m pytest --ds=cms.envs.test openedx/core/lib/blockstore_api/ openedx/core/djangolib/tests/test_blockstore_cache.py openedx/core/djangoapps/content_libraries/tests/
```
